lib/msal-core/src/Configuration.ts

```diff
@@ -0,0 +1,181 @@
+  state?: string;
+  navigateToLoginRequestUrl?: boolean;
+};
+
+/**
+ * Cache Options
+ *
+ * cacheLocation            - Used to specify the cacheLocation user wants to set: Valid values are "localStorage" and "sessionStorage"
+ * storeAuthStateInCookie   - If set, the library will store the auth request state required for validation of the auth flows in the browser cookies.
+ *                              By default this flag is set to false.
+ */
+export type CacheOptions = {
+  cacheLocation?: CacheLocation;
+  storeAuthStateInCookie?: boolean;
+};
+
+/**
+ * Library Specific Options
+ *
+ * logger                       - Used to initialize the Logger object; TODO: Expand on logger details or link to the documentation on logger
+ * loadFrameTimeout             - maximum time the library should wait for a frame to load
+ * tokenRenewalOffsetSeconds    - sets the window of offset needed to renew the token before expiry
+ *
+ */
+export type SystemOptions = {
+  logger?: Logger;
+  loadFrameTimeout?: number;
+  tokenRenewalOffsetSeconds?: number;
+};
+
+/**
+ * App/Framework specific environment Support
+ *
+ * isAngular                - flag set to determine if it is Angular Framework. Used to broadcast tokens. TODO: detangle this dependency from core.
+ * unprotectedResources     - Array of URI's which are unprotected resources. MSAL will not attach a token to outgoing requests that have these URI. Defaults to 'null'.
+ * protectedResourceMap     - This is mapping of resources to scopes used by MSAL for automatically attaching access tokens in web API calls.
+ *                              A single access token is obtained for the resource. So you can map a specific resource path as follows:
+ *                              {"https://graph.microsoft.com/v1.0/me", ["user.read"]},
+ *                              or the app URL of the resource as: {"https://graph.microsoft.com/", ["user.read", "mail.send"]}.
+ *                              This is required for CORS calls.
+ *
+ */
+export type FrameworkOptions = {
+  isAngular?: boolean;
+  unprotectedResources?: Array<string>;
+  protectedResourceMap?: Map<string, Array<string>>;
+};
+
+/**
+ * Configuration Object
+ */
+export type Configuration = {
+  auth: AuthOptions,
+  cache?: CacheOptions,
+  system?: SystemOptions,
+  framework?: FrameworkOptions
+};
+
+/**
+ * Function to set the default options when not explicitly set
+ *
+ * @param TAuthOptions
+ * @param TCacheOptions
+ * @param TSystemOptions
+ * @param TFrameworkOptions
+ *
+ * @returns TConfiguration object
+ */
+
+// destructure with default settings
+export function buildConfiguration(
+  {
+    clientId = "",
+    authority = null,
+    validateAuthority = true,
+    redirectUri = () => window.location.href.split("?")[0].split("#")[0],
+    postLogoutRedirectUri = () => window.location.href.split("?")[0].split("#")[0],
+    state = "",
+    navigateToLoginRequestUrl = true
+  }: AuthOptions = { clientId },
```

Comment:
not sure I full understand this, making sure
define the defaults for authOptions destructure so you can use
what does = { clientId } do?

Reply:
Since clientId is a mandatory field, the User should definitely pass the client Id. This code ensures that.

Reply:
I think I understand, but I also think that the = { clientId } might be unencessary 
its essentially destructuring the input and making all of the variables have defaults that are also available to the function to use. It is not clear to me what the = { clientId } does. I don't think it makes it available to the fn because it already is

Reply:
Yeah I was just being uniform across my declarations. I tested removing the initializations to empty objects and it seems to work fine. I also need to fine tune the initialization of Config object for better flexibility to the developer, that can be experimented during samples. 

I added the change now removing the {clientId}. Please review and let me know.